# TASK #1. DEFINE A PANDAS SERIES

In [1]:
# Pandas is a data manipulation and analysis tool that is built on Numpy.
# Pandas uses a data structure known as DataFrame (think of it as Microsoft excel in Python). 
# DataFrames empower programmers to store and manipulate data in a tabular fashion (rows and columns).
# Series Vs. DataFrame? Series is considered a single column of a DataFrame.
import pandas as pd


In [2]:
# Let's define a Python list that contains 5 crypto currencies 
crypto_list = ["BTC", "XRP", "LTC", "ADA", "ETH",]


In [3]:
# Let's confirm the Datatype
type(crypto_list)


list

In [4]:
# Let's create a one dimensional Pandas "series" 
# Let's use Pandas Constructor Method to create a series from a Python list
# Note that series is formed of data and associated index (numeric index has been automatically generated) 
# Check Pandas Documentation for More information: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html#pandas.Series
# Object datatype is used for text data (String)
crypto_series = pd.Series(data=crypto_list)
crypto_series


0    BTC
1    XRP
2    LTC
3    ADA
4    ETH
dtype: object

In [5]:
# Let's confirm the Pandas Series Datatype
type(crypto_series)


pandas.core.series.Series

In [6]:
# Let's define another Pandas Series that contains numeric values (crypto prices) instead of text data
# Note that we have int64 datatype which means it's integer stored in 64 bits in memory
crypto_prices_series = pd.Series(data=[2000, 500, 2000, 20, 50])
crypto_prices_series


0    2000
1     500
2    2000
3      20
4      50
dtype: int64

**MINI CHALLENGE #1:**
- **Define a Pandas Series named "my_series" that contains your top 3 favourite stocks. Confirm the datatype of "my_series"**

In [7]:
favourite_stocks = ["IBM", "Nvidia", "Microsoft"]
my_series = pd.Series(data=favourite_stocks)
my_series


0          IBM
1       Nvidia
2    Microsoft
dtype: object

In [8]:
type(my_series)

pandas.core.series.Series

# TASK #2. DEFINE A PANDAS SERIES WITH CUSTOM INDEX

In [9]:
# Let's define a Python list that contains 5 Crypto currencies
crypto_list = ["BTC", "XRP", "LTC", "ADA", "ETH",]
crypto_list


['BTC', 'XRP', 'LTC', 'ADA', 'ETH']

In [10]:
# Let's define a python list as shown below. This python list will be used for the Series index:
crypto_labels = ["crypto#1", "crypto#2", "crypto#3", "crypto#4", "crypto#5"]
crypto_labels


['crypto#1', 'crypto#2', 'crypto#3', 'crypto#4', 'crypto#5']

In [11]:
# Let's create a one dimensional Pandas "series" 
# Let's use Pandas Constructor Method to create a series from a Python list
# Note that this series is formed of data and associated labels 
crypto_series = pd.Series(data=crypto_list, index=crypto_labels)


In [12]:
# Let's view the series
crypto_series

crypto#1    BTC
crypto#2    XRP
crypto#3    LTC
crypto#4    ADA
crypto#5    ETH
dtype: object

In [13]:
# Let's obtain the datatype
type(crypto_series)


pandas.core.series.Series

**MINI CHALLENGE #2:**
- **Define a Pandas Series named "my_series" that contains your top 3 favourite stocks. Instead of using default numeric indexes (similar to mini challenge #1), use the following indexes "stock #1", "stock #2", and "stock #3"**

In [14]:
favourite_stocks = ["IBM", "Nvidia", "Microsoft"]
stocks_labels = ["stock#1", "stock#2", "stock#3"]
my_series = pd.Series(data=favourite_stocks, index=stocks_labels)
my_series


stock#1          IBM
stock#2       Nvidia
stock#3    Microsoft
dtype: object

# TASK #3. DEFINE A PANDAS SERIES FROM A DICTIONARY

In [15]:
# A Dictionary consists of a collection of key-value pairs. Each key-value pair maps the key to its corresponding value.
# Keys are unique within a dictionary while values may not be. 
# List elements are accessed by their position in the list, via indexing while Dictionary elements are accessed via keys
# Define a dictionary named "my_dict" using key-value pairs
my_dict = {"Employee ID": 1,
          "Employee Name": "Steve",
          "Salary $": 2000,
          "Years with Company": 10}


In [16]:
# Show the dictionary
my_dict


{'Employee ID': 1,
 'Employee Name': 'Steve',
 'Salary $': 2000,
 'Years with Company': 10}

In [17]:
# Confirm the dictionary datatype
type(my_dict)


dict

In [18]:
# Let's define a Pandas Series Using the dictionary
employee_series = pd.Series(my_dict)
employee_series


Employee ID               1
Employee Name         Steve
Salary $               2000
Years with Company       10
dtype: object

**MINI CHALLENGE #3:**
- **Create a Pandas Series from a dictionary with 3 of your favourite stocks and their corresponding prices** 

In [19]:
favourite_stocks_dict = {"IBM": 200, "Nvidia": 2100, "Microsoft": 950}
stocks_series = pd.Series(favourite_stocks_dict)
stocks_series


IBM           200
Nvidia       2100
Microsoft     950
dtype: int64

# TASK #4. PANDAS ATTRIBUTES

In [20]:
# Attributes/Properties: do not use parantheses "()" and are used to get Pandas Series Properties. Ex: my_series.values, my_series.shape
# Methods: use parantheses "()" and might include arguments and they actually alter/change the Pandas Series. Ex: my_series.tail(), my_series.head(), my_series.drop_duplicates()
# Indexers: use square brackets "[]" and are used to access specific elements in a Pandas Series or DataFrame. Ex: my_series.loc[], my_series.iloc[]

# Let's redefine a Pandas Series containing our favourite 5 cryptos 
crypto_list = ["BTC", "XRP", "LTC", "ADA", "ETH",]
crypto_series = pd.Series(data=crypto_list)
crypto_series


0    BTC
1    XRP
2    LTC
3    ADA
4    ETH
dtype: object

In [21]:
# ".Values" attribute is used to return Series as ndarray depending on its dtype
# Check this for more information: https://pandas.pydata.org/docs/reference/api/pandas.Series.values.html#pandas.Series.values
crypto_series.values


array(['BTC', 'XRP', 'LTC', 'ADA', 'ETH'], dtype=object)

In [22]:
# index is used to return the index (axis labels) of the Series
crypto_series.index


RangeIndex(start=0, stop=5, step=1)

In [23]:
# dtype is used to return the datatype of the Series ('O' stands for 'object' datatype)
crypto_series.dtype

dtype('O')

In [24]:
# Check if all elements are unique or not
crypto_series.is_unique


True

In [25]:
# Check the shape of the Series
# note that a Series is one dimensional
crypto_series.shape


(5,)

**MINI CHALLENGE #4:** 
- **What is the size of the Pandas Series? (External Research for the proper attribute is Required)**

In [26]:
crypto_series.size


5

# TASK #5. PANDAS METHODS

In [27]:
# Methods have parentheses and they actually alter/change the Pandas Series
# Methods: use parantheses "()" and might include arguments. Ex: my_series.tail(), my_series.head(), my_series.drop_duplicates()

# Let's define another Pandas Series that contains numeric values (crypto prices) instead of text data
# Note that we have int64 datatype which means it contains integer values stored in 64 bits in memory
crypto_prices = pd.Series(data=[400, 500, 1500, 20, 70])
crypto_prices


0     400
1     500
2    1500
3      20
4      70
dtype: int64

In [28]:
# Let's obtain the sum of all elements in the Pandas Series
crypto_prices.sum()


2490

In [29]:
# Let's obtain the multiplication of all elements in the Pandas Series
crypto_prices.product()


420000000000

In [30]:
# Let's obtain the average
crypto_prices.mean()


498.0

In [32]:
# Let's show the first couple of elements in the Pandas Series
crypto_prices.head(2)


0    400
1    500
dtype: int64

In [33]:
# Note that head creates a new dataframe 
new_crypto_prices = crypto_prices.head(3)
new_crypto_prices


0     400
1     500
2    1500
dtype: int64

**MINI CHALLENGE #5:** 
- **Show the last 2 rows in the Pandas Series (External Research is Required)** 
- **How many bytes does this Pandas Series consume in memory? (External Research is Required)**

In [34]:
crypto_prices.tail(2)


3    20
4    70
dtype: int64

In [35]:
crypto_series.memory_usage()


172

# TASK #6. IMPORT CSV DATA (1-D) USING PANDAS

In [34]:
# Pandas read_csv is used to read a csv file and store data in a DataFrame by default (DataFrames will be covered shortly!)
# Use Squeeze to convert it into a Pandas Series (One-dimensional)
# Notice that no foramtting exists when a Series is plotted

0         457.334015
1         424.440002
2         394.795990
3         408.903992
4         398.821014
            ...     
2380    55950.746090
2381    57750.199220
2382    58917.691410
2383    58918.832030
2384    59095.808590
Name: BTC-USD Price, Length: 2385, dtype: float64

**MINI CHALLENGE #6:**
- **Set Squeeze = False and rerun the cell, what do you notice? Use Type to compare both outputs**

BTC-USD Price
0        457.334015
1        424.440002
2        394.795990
3        408.903992
4        398.821014
...             ...
2380   55950.746090
2381   57750.199220
2382   58917.691410
2383   58918.832030
2384   59095.808590

[2385 rows x 1 columns]

BTC_price_series type: <class 'pandas.core.series.Series'>
BTC_price_df type: <class 'pandas.core.frame.DataFrame'>


# TASK #7. PANDAS BUILT-IN FUNCTIONS

In [38]:
# Pandas works great with pre-existing python functions 
# You don't have to play with pandas methods and directly leverage Python functions
# Check Python built-in functions here: https://docs.python.org/3/library/functions.html

0         457.334015
1         424.440002
2         394.795990
3         408.903992
4         398.821014
            ...     
2380    55950.746090
2381    57750.199220
2382    58917.691410
2383    58918.832030
2384    59095.808590
Name: BTC-USD Price, Length: 2385, dtype: float64

In [39]:
# Obtain the Data Type of the Pandas Series

pandas.core.series.Series

In [40]:
# Obtain the length of the Pandas Series

2385

In [41]:
# Obtain the maximum value of the Pandas Series

61243.08594

In [42]:
# Obtain the minimum value of the Pandas Series

178.1029968

**MINI CHALLENGE #7:**
- **Given the following Pandas Series, convert all positive values to negative using python built-in functions**
- **Obtain only unique values (ie: Remove duplicates) using python built-in functions**
- my_series = pd.Series(data = [-10, 100, -30, 50, 100])


0     10
1    100
2    -30
3     50
4    100
dtype: int64

In [53]:
# Convert all positive values to negative

0    -10
1   -100
2    -30
3    -50
4   -100
dtype: int64

{-30, 10, 50, 100}

# TASK #8. SORTING PANDAS SERIES

In [55]:
# Let's import CSV data as follows:

0         457.334015
1         424.440002
2         394.795990
3         408.903992
4         398.821014
            ...     
2380    55950.746090
2381    57750.199220
2382    58917.691410
2383    58918.832030
2384    59095.808590
Name: BTC-USD Price, Length: 2385, dtype: float64

In [56]:
# You can sort the values in the dataframe as follows

119       178.102997
122       199.259995
121       208.097000
120       209.843994
123       210.339004
            ...     
2382    58917.691410
2383    58918.832030
2384    59095.808590
2366    59302.316410
2365    61243.085940
Name: BTC-USD Price, Length: 2385, dtype: float64

In [57]:
# Let's view Pandas Series again after sorting, Note that nothing changed in memory! you have to make sure that inplace is set to True

0         457.334015
1         424.440002
2         394.795990
3         408.903992
4         398.821014
            ...     
2380    55950.746090
2381    57750.199220
2382    58917.691410
2383    58918.832030
2384    59095.808590
Name: BTC-USD Price, Length: 2385, dtype: float64

In [58]:
# Set inplace = True to ensure that change has taken place in memory 

In [59]:
# Note that now the change (ordering) took place 

119       178.102997
122       199.259995
121       208.097000
120       209.843994
123       210.339004
            ...     
2382    58917.691410
2383    58918.832030
2384    59095.808590
2366    59302.316410
2365    61243.085940
Name: BTC-USD Price, Length: 2385, dtype: float64

In [61]:
# Notice that the indexes are now changed 
# You can also sort by index (revert back to the original Pandas Series) as follows: 

0         457.334015
1         424.440002
2         394.795990
3         408.903992
4         398.821014
            ...     
2380    55950.746090
2381    57750.199220
2382    58917.691410
2383    58918.832030
2384    59095.808590
Name: BTC-USD Price, Length: 2385, dtype: float64

**MINI CHALLENGE #8:**
- **Sort the BTC_price_series values in a decending order instead. Make sure to update values in-memory.**

2365    61243.085940
2366    59302.316410
2384    59095.808590
2383    58918.832030
2382    58917.691410
            ...     
123       210.339004
120       209.843994
121       208.097000
122       199.259995
119       178.102997
Name: BTC-USD Price, Length: 2385, dtype: float64

# TASK #9. PERFORM MATH OPERATIONS ON PANDAS SERIES

In [41]:
# Let's import CSV data as follows:


0         457.334015
1         424.440002
2         394.795990
3         408.903992
4         398.821014
            ...     
2380    55950.746090
2381    57750.199220
2382    58917.691410
2383    58918.832030
2384    59095.808590
Name: BTC-USD Price, Length: 2385, dtype: float64

In [42]:
# Apply Sum Method on Pandas Series


15435379.738852698

In [43]:
# Apply count Method on Pandas Series


2385

In [44]:
# Obtain the maximum value


61243.08594

In [45]:
# Obtain the minimum value


178.1029968

In [46]:
# My favourite: Describe! 
# Describe is used to obtain all statistical information in one place 


count     2385.000000
mean      6471.857333
std       9289.022505
min        178.102997
25%        454.618988
50%       4076.632568
75%       8864.766602
max      61243.085940
Name: BTC-USD Price, dtype: float64

**MINI CHALLENGE #9:**
- **Obtain the average price of the BTC_price_series using two different methods**

# TASK #10. CHECK IF A GIVEN ELEMENT EXISTS IN A PANDAS SERIES

In [47]:
# Let's import CSV data as follows:


0         457.334015
1         424.440002
2         394.795990
3         408.903992
4         398.821014
            ...     
2380    55950.746090
2381    57750.199220
2382    58917.691410
2383    58918.832030
2384    59095.808590
Name: BTC-USD Price, Length: 2385, dtype: float64

In [48]:
# Check if a given number exists in a Pandas Series values
# Returns a boolean "True" or "False"


False

In [49]:
# Check if a given number exists in a Pandas Series index


True

In [50]:
# Note that by default 'in' will search in Pandas index and not values


True

**MINI CHALLENGE #10:**
- **Check if the stock price 399 exists in the BTC_price_series Pandas Series or not**
- **Round stock prices to the nearest integer and check again**

# EXCELLENT JOB!